stan/math/opencl/kernel_generator/block.hpp

```diff
@@ -0,0 +1,225 @@
+    res.args = "int " + var_name + "_i, int " + var_name + "_j, ";
+    return res;
+  }
+
+  /**
+   * Sets offset of block to indices of the argument expression
+   * @param[in, out] i row index
+   * @param[in, out] j column index
+   */
+  inline void modify_argument_indices(std::string& i, std::string& j) const {
+    i = "(" + i + " + " + var_name + "_i)";
+    j = "(" + j + " + " + var_name + "_j)";
+  }
+
+  /**
+   * generates kernel code for this and nested expressions if this expression
+   * appears on the left hand side of an assignment.
+   * @param[in,out] generated set of already generated operation_cls
+   * @param ng name generator for this kernel
+   * @param i row index variable name
+   * @param j column index variable name
+   * @return part of kernel with code for this and nested expressions
+   */
+  inline kernel_parts generate_lhs(const std::string& i, const std::string& j,
+                                   const std::string& var_name_arg) const {
+    kernel_parts res;
+    res.args = "int " + var_name + "_i, int " + var_name + "_j, ";
+    return res;
+  }
+
+  /**
+   * Sets kernel arguments for this and nested expressions.
+   * @param[in,out] generated set of expressions that already set their kernel
+   * arguments
+   * @param kernel kernel to set arguments on
+   * @param[in,out] arg_num consecutive number of the first argument to set.
+   * This is incremented for each argument set by this function.
+   */
+  inline void set_args(std::set<const operation_cl_base*>& generated,
+                       cl::Kernel& kernel, int& arg_num) const {
+    if (generated.count(this) == 0) {
+      generated.insert(this);
+      std::get<0>(arguments_).set_args(generated, kernel, arg_num);
+      kernel.setArg(arg_num++, start_row_);
+      kernel.setArg(arg_num++, start_col_);
+    }
+  }
+
+  /**
+   * View of a matrix that would be the result of evaluating this expression.
+   * @return view
+   */
+  inline matrix_cl_view view() const { return std::get<0>(arguments_).view(); }
+
+  /**
+   * Number of rows of a matrix that would be the result of evaluating this
+   * expression.
+   * @return number of rows
+   */
+  inline int rows() const { return rows_; }
+
+  /**
+   * Number of columns of a matrix that would be the result of evaluating this
+   * expression.
+   * @return number of columns
+   */
+  inline int cols() const { return cols_; }
+
+  /**
+   * Sets view of the underlying matrix depending on which part is written.
+   * @param top_diagonal Index of the top sub- or super- diagonal written with
+   * nonzero elements.
+   * @param bottom_diagonal Index of the top sub- or super- diagonal written
+   * with nonzero elements.
+   * @param top_zero_diagonal Index of the top sub- or super- diagonal written
+   * with zeros if it ie more extreme than \c top_diagonal. Otherwise it should
+   * be set to equal value as \c top_diagonal.
+   * @param bottom_zero_diagonal Index of the top sub- or super- diagonal
```

Comment:
It would be nice to have the same order of arguments in the comments as in the signatures.